### ISSUE TITLE
Plugin does not use the right SCM reliably

### ISSUE THREAD

**Issue by Xaseron** [jenkinsci/last-changes-plugin#25]
My pipeline code is in a svn repository and the project ist stored in a git repository.
I execute the lastChanges step right after the checkout of the git repository.
Sometimes it fails because of the exception listed below.  It looks to me, that the order of items from `SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(projectAction.getProject()).getSCMs()`  is random and sometimes the plugin tries to convert the git commit hash into an svn revsion. 

```
java.lang.NumberFormatException: For input string: "784add6adb524b84bb3287607a9fded9b34af16c"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
        at java.lang.Long.parseLong(Long.java:589)
        at java.lang.Long.parseLong(Long.java:631)
        at com.github.jenkins.lastchanges.LastChangesPublisher.perform(LastChangesPublisher.java:181)
        at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
        at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
        at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:49)
        at hudson.security.ACL.impersonate(ACL.java:260)
        at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:46)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by rmpestano** [comment]
Hi @Xaseron,

The main problem is how last changes deal with SVN repo, it needs to access it remotelly where in GIT we just access the local .git repo. 

I will try (one more time) to get SVN revision info from local repository and if it works then we can get ride of `SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(projectAction.getProject()).getSCMs()` and rely only on checked out/cloned local repo.

**Comment by rmpestano** [comment]
Hi @Xaseron,

I think I got it working now, can you try with attached plugin? You'll need the extension to `.hpi` before installing into your Jenkins instance:
[last-changes.zip](https://github.com/jenkinsci/last-changes-plugin/files/1434304/last-changes.zip)

**Comment by Xaseron** [comment]
i did a couple of test runs and i couldn't reproduce the bug 👍

**Comment by rmpestano** [comment]
Great, that's great news. There should be a release as soon #26 is implemented.

Thank you

**Comment by Xaseron** [comment]
I've tested it with only a svn repository and i get a E170001: Authentication required for '<repository> Authorization Realm'

How can i supply credentials for the svn lastChanges?

**Comment by rmpestano** [comment]
Hi, authentication should not be required anymore as we are reading changes from local repo(.svn) folder.


I will do some tests on a private svn repo.

Also do you have the complete stacktrace?

For GIT repo It works or your git repo is public'?

**Comment by Xaseron** [comment]
My git repo is private too.

```
java.io.IOException: Failed to persist /var/lib/jenkins/jenkins1/jobs/artf445380_Code-DIFF-Step/builds/48/program.dat
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:472)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:434)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgramIfPossible(CpsThreadGroup.java:422)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:362)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:82)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:242)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:230)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.NotSerializableException: org.tmatesoft.svn.core.SVNURL
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:860)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:778)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverObjectOutputStream.defaultWriteObject(RiverObjectOutputStream.java:165)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.GeneratedMethodAccessor22.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverObjectOutputStream.defaultWriteObject(RiverObjectOutputStream.java:165)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.GeneratedMethodAccessor22.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.BlockMarshaller.doWriteObject(BlockMarshaller.java:65)
	at org.jboss.marshalling.river.BlockMarshaller.writeObject(BlockMarshaller.java:56)
	at org.jboss.marshalling.MarshallerObjectOutputStream.writeObjectOverride(MarshallerObjectOutputStream.java:50)
	at org.jboss.marshalling.river.RiverObjectOutputStream.writeObjectOverride(RiverObjectOutputStream.java:179)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:344)
	at java.util.HashMap.internalWriteEntries(HashMap.java:1777)
	at java.util.HashMap.writeObject(HashMap.java:1354)
	at sun.reflect.GeneratedMethodAccessor29.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.BlockMarshaller.doWriteObject(BlockMarshaller.java:65)
	at org.jboss.marshalling.river.BlockMarshaller.writeObject(BlockMarshaller.java:56)
	at org.jboss.marshalling.MarshallerObjectOutputStream.writeObjectOverride(MarshallerObjectOutputStream.java:50)
	at org.jboss.marshalling.river.RiverObjectOutputStream.writeObjectOverride(RiverObjectOutputStream.java:179)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:344)
	at java.util.TreeMap.writeObject(TreeMap.java:2434)
	at sun.reflect.GeneratedMethodAccessor71.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.AbstractObjectOutput.writeObject(AbstractObjectOutput.java:58)
	at org.jboss.marshalling.AbstractMarshaller.writeObject(AbstractMarshaller.java:111)
	at org.jenkinsci.plugins.workflow.support.pickles.serialization.RiverWriter.writeObject(RiverWriter.java:140)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:458)
	... 15 more
Caused by: an exception which occurred:
	in field myObjects
	in field myChildErrorMessage
	in field myErrorMessage
	in field cause
	in field cause
	in field locals
	in field parent
	in field parent
	in field caller
	in field parent
	in field parent
	in field parent
	in field caller
	in field parent
	in field parent
	in field caller
	in field e
	in field program
	in field threads
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@61a2655f
```

**Comment by rmpestano** [comment]
Hi, I don't see last changes on the stacktrace, needed to know which part is raising the exception.  Are you running on a slave?

**Comment by Xaseron** [comment]
Sorry, i've copied the wrong exception.  

Here is the right one. (i anonymised the svn repository with <secret>)
```
java.lang.RuntimeException: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jenkins1/jobs/artf445380_Code-DIFF-Step/builds/52 due to following error: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jenkins1/jobs/artf445380_Code-DIFF-Step/builds/52 due to following error: svn: E170001: Authentication required for '<secret> Authorization Realm' - org.tmatesoft.svn.core.SVNAuthenticationException: svn: E170001: Authentication required for '<secret> Authorization Realm'
	at com.github.jenkins.lastchanges.impl.SvnLastChanges.changesOf(SvnLastChanges.java:59)
	at com.github.jenkins.lastchanges.LastChangesPublisher.perform(LastChangesPublisher.java:182)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:49)
	at hudson.security.ACL.impersonate(ACL.java:260)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:46)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jenkins1/jobs/artf445380_Code-DIFF-Step/builds/52 due to following error: svn: E170001: Authentication required for '<secret> Authorization Realm'
	at com.github.jenkins.lastchanges.impl.SvnLastChanges.changesOf(SvnLastChanges.java:95)
	at com.github.jenkins.lastchanges.impl.SvnLastChanges.changesOf(SvnLastChanges.java:57)
	... 11 more
Caused by: org.tmatesoft.svn.core.SVNAuthenticationException: svn: E170001: Authentication required for '<secret> Authorization Realm'
	at org.tmatesoft.svn.core.internal.wc.SVNErrorManager.authenticationFailed(SVNErrorManager.java:53)
	at org.tmatesoft.svn.core.internal.wc.SVNErrorManager.authenticationFailed(SVNErrorManager.java:47)
	at org.tmatesoft.svn.core.internal.wc.DefaultSVNAuthenticationManager.getFirstAuthentication(DefaultSVNAuthenticationManager.java:201)
	at org.tmatesoft.svn.core.internal.io.dav.http.HTTPConnection.request(HTTPConnection.java:687)
	at org.tmatesoft.svn.core.internal.io.dav.http.HTTPConnection.request(HTTPConnection.java:352)
	at org.tmatesoft.svn.core.internal.io.dav.http.HTTPConnection.request(HTTPConnection.java:340)
	at org.tmatesoft.svn.core.internal.io.dav.DAVConnection.performHttpRequest(DAVConnection.java:888)
	at org.tmatesoft.svn.core.internal.io.dav.DAVConnection.exchangeCapabilities(DAVConnection.java:693)
	at org.tmatesoft.svn.core.internal.io.dav.DAVConnection.open(DAVConnection.java:113)
	at org.tmatesoft.svn.core.internal.io.dav.DAVRepository.openConnection(DAVRepository.java:1030)
	at org.tmatesoft.svn.core.internal.io.dav.DAVRepository.getLatestRevision(DAVRepository.java:164)
	at org.tmatesoft.svn.core.internal.wc16.SVNBasicDelegate.getRevisionNumber(SVNBasicDelegate.java:480)
	at org.tmatesoft.svn.core.internal.wc16.SVNBasicDelegate.getRevisionNumber(SVNBasicDelegate.java:464)
	at org.tmatesoft.svn.core.internal.wc16.SVNDiffClient16.doDiffURLURL(SVNDiffClient16.java:3091)
	at org.tmatesoft.svn.core.internal.wc16.SVNDiffClient16.doDiff(SVNDiffClient16.java:1067)
	at org.tmatesoft.svn.core.internal.wc2.old.SvnOldDiff.run(SvnOldDiff.java:61)
	at org.tmatesoft.svn.core.internal.wc2.old.SvnOldDiff.run(SvnOldDiff.java:15)
	at org.tmatesoft.svn.core.internal.wc2.SvnOperationRunner.run(SvnOperationRunner.java:21)
	at org.tmatesoft.svn.core.wc2.SvnOperationFactory.run(SvnOperationFactory.java:1235)
	at org.tmatesoft.svn.core.wc2.SvnOperation.run(SvnOperation.java:294)
	at com.github.jenkins.lastchanges.impl.SvnLastChanges.changesOf(SvnLastChanges.java:87)
	... 12 more
```

**Comment by rmpestano** [comment]
Is the job running on the same machine that checked out the code? 

I ask because by default svn tries to get credentials from `~/.subversion` folder.

**Comment by rmpestano** [comment]
Hi again, can you try with the following plugin?

[last-changes.zip](https://github.com/jenkinsci/last-changes-plugin/files/1441236/last-changes.zip)


Don't forget to rename de .zip to .hpi

**Comment by Xaseron** [comment]
It doesn't crash anymore. But i can determine the previous revision and i can't set the parameter.
`10:18:11 Last changes from revision null to 3928 published successfully!`

**Comment by rmpestano** [comment]
Hi @Xaseron, I've just fixed the problem of null (issue #30) and it is available in v2.2.
I'll close this issue and if you find any problem related to this just reopen it.

**Comment by jaredrsowers** [comment]
I am seeing the same exact error running 2.6.8. Everything was working fine for a couple of weeks until today, and it is now failing with:

> ERROR: Last Changes NOT published due to the following error: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jobs/ZZZ/branches/branches-zzz.7sqi9n/builds/42 due to following error: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jobs/ZZZ/branches/branches-zzz.7sqi9n/builds/42 due to following error: svn: E170001: Authentication required for '<https://svn.zzz.com:443> VisualSVN Server' - org.tmatesoft.svn.core.SVNAuthenticationException: svn: E170001: Authentication required for '<https://svn.zzz.com:443> VisualSVN Server' - java.lang.RuntimeException: Could not retrieve last changes of svn repository located at /var/lib/jenkins/jobs/ZZZ/branches/branches-zzz.7sqi9n/builds/42 due to following error: svn: E170001: Authentication required for '<https://svn.zzz.com:443> VisualSVN Server'

**Comment by jaredrsowers** [comment]
One thing that has changed is I started using a Global Pipeline Library from a git repo. Just removing the library doesn't show any change in the issue, but if I wait several hours after having removed the library the issue seems to go away, and on the inverse, when I add the library again it will continue to work for several hours and then the error comes back.

Honestly I am unsure if this is actually related in any way, but it is the only thing that is new...

**Comment by rmpestano** [comment]
Hi, can you share how you checkout the code into the workspace on your pipeline or shared library?

**Comment by jaredrsowers** [comment]
The Global Pipeline Library:

![image](https://user-images.githubusercontent.com/16364079/48651937-f42eef80-e9b1-11e8-8bb7-3f6d52715168.png)

Then the pipeline is a Multibranch Pipeline:

![image](https://user-images.githubusercontent.com/16364079/48652022-5ee02b00-e9b2-11e8-89a9-0d031f914114.png)

And the pipeline is setup as follows:

```
pipeline {

    options {
        skipDefaultCheckout true
    }
```
```

            stages {
                stage('Checkout') {
                    steps {
                        script {
                            def scmVars = checkout scm

                            SVN_REVISION = scmVars.SVN_REVISION
                            SVN_URL = scmVars.SVN_URL
                        }
                    }
                }
                stage('Initialize') {
                    steps {
                        script {

                            isDev = "${BRANCH_NAME}".startsWith("branches/zzzzz")


                            specFiles = findFiles(glob: 'build/*.spec')

                            if (isDev) {
                                def devSpecFiles = findFiles(glob: 'build/dev/*.spec')

                                specFiles = specFiles.plus(devSpecFiles)
                            }

                            def publisher = LastChanges.getLastChangesPublisher "LAST_SUCCESSFUL_BUILD", "SIDE", "LINE", true, true, "", "", "", "", ""
                            publisher.publishLastChanges()
                            def changes = publisher.getLastChanges()

                            def hasChanges = false

                            // added this check because changes is null when it gets the auth error
                            if (changes != null) {

                                def diff = changes.getDiff()

                                for (commit in changes.getCommits()) {
                                    def commitInfo = commit.getCommitInfo()

                                    svnChangelog = "${svnChangelog}<hr />r${commitInfo.getCommitId()}  |  <strong>${commitInfo.getCommitterName()}</strong>  |  ${commitInfo.getCommitDate()}<br />------------------------------------------------------------------------<br />${commitInfo.getCommitMessage()}"

                                    if (hasChanges) {
                                        if (culpritNames.contains(commitInfo.getCommitterName()) == false) {
                                            culpritNames = "${culpritNames} and ${commitInfo.getCommitterName()}"
                                        }
                                    } else {
                                        culpritNames = "${commitInfo.getCommitterName()}"
                                    }

                                    if (errorCulpritNames.contains(commitInfo.getCommitterName()) == false) {
                                        errorCulpritNames = "${errorCulpritNames} and <span style='color: red;'>${commitInfo.getCommitterName()}</span>"
                                    }

                                    hasChanges = true;
                                }
                            }

                            if (hasChanges) {
                                writeFile file: 'build.diff', text: diff
                            } else {
                                svnChangelog = '<hr />No Changes'
                                culpritNames = 'Someone'
                            }

                        }


```

**Comment by rmpestano** [comment]
I didn't understand how your checkout happens, you don't provide any credentials?

**Comment by rmpestano** [comment]
> I didn't understand how your checkout happens, you don't provide any credentials?
Now I see your second image. As you're using multibranch the problem [may be here](https://github.com/jenkinsci/last-changes-plugin/blob/cd15d8b6ef762726644a13a76aa376aab615097e/src/main/java/com/github/jenkins/lastchanges/LastChangesPublisher.java#L162) because we're getting the auth manager from first scm always.

I don't know how to solve neither to reproduce as I don't work with multibranch pipelines, we have other issues (#42) so this may be the cause. 

Have you faced the issue in a non multibranch pipeline as well?

**Comment by jaredrsowers** [comment]
This setup had been working without issue until yesterday. 

The day before yesterday, in the evening, I added the Global Pipeline Library from git, and everything continued to work, then yesterday morning this error showed up. In my trouble shooting I deleted the Global Pipeline Library to see if it would help, and it did not. Then yesterday evening around 7:00 it magically started working again. Then today I added the Global Pipeline Library back in and it continued to work for several hours and now the error is back...

My next step is to move the Global Pipeline Library to SVN, and something tells me that the error is going to go away...
